Frontend/CO.CDP.OrganisationApp/Pages/Forms/_FormElementUrlInput.cshtml

```diff
@@ -0,0 +1,97 @@
+@model FormElementUrlInputModel
+
+@{
+    var textInputHasError = ((TagBuilder)Html.ValidationMessageFor(m => m.TextInput)).HasInnerHtml;
+    var selectOptionHasError = ((TagBuilder)Html.ValidationMessageFor(m => m.HasValue)).HasInnerHtml;
+    string textInputAriaDescribedby = "";
+
+    void RenderHeading()
+    {
+        @if (!string.IsNullOrWhiteSpace(Model.Heading))
+        {
+            <h1 class="govuk-label-wrapper">
+                <label class="govuk-label govuk-label--l" for="TextInput">@Model.Heading</label>
+
+                @if (!string.IsNullOrWhiteSpace(Model.Caption))
+                {
+                    <span class="govuk-caption-l govuk-!-margin-bottom-3">@Model.Caption</span>
+                }
+            </h1>
+        }
+
+        @if (!string.IsNullOrWhiteSpace(Model.Description))
+        {
+            textInputAriaDescribedby += "TextInput-description";
+
+            <div id="TextInput-description">
+                @Html.Raw(Model.Description)
+            </div>
+        }
+
+        @if (selectOptionHasError)
+        {
+            <p class="govuk-error-message">
+                <span class="govuk-visually-hidden">Error:</span>
+                @Html.ValidationMessageFor(m => m.HasValue)
+            </p>
+        }
+    }
+
+    void RenderField()
+    {
+        if (textInputHasError)
+        {
+            textInputAriaDescribedby += " TextInput-error";
+
+            <p class="govuk-error-message" id="TextInput-error">
+                <span class="govuk-visually-hidden">Error:</span>
+                @Html.ValidationMessageFor(m => m.TextInput)
+            </p>
+        }
+
+        <input class="govuk-input govuk-!-width-two-thirds" id="TextInput" name="TextInput" type="text" value="@Model.TextInput"
+        @(string.IsNullOrWhiteSpace(textInputAriaDescribedby) ? "" : Html.Raw($"aria-describedby=\"{textInputAriaDescribedby}\""))>
+    }
+}
+
+<div class="govuk-form-group @((Model.IsRequired && textInputHasError) || selectOptionHasError ? "govuk-form-group--error" : "")">
+
+    @if (Model.IsRequired)
+    {
+        RenderHeading();
+        RenderField();
+    }
+    else
+    {
+        <fieldset class="govuk-fieldset">
+            <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
+                @{
+                    RenderHeading();
```

Comment:
Closer, but still not quite right. The RenderHeading method has too much in it. The legend should be this and nothing more:

```
<legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
    <h1 class="govuk-fieldset__heading">
      @Model.Heading
    </h1>
  </legend>

  ... descriptions, errors and everything else goes below ...
```

I feel like in an attempt to keep this DRY, it's steering away from the right markup in each instance and it would be better to repeat yourself slightly.

Reply:
Very well pointed @andymantell